### DISCUSSION TITLE
20.4.0 and 21.0

### DISCUSSION THREAD

**Discussion by fballiano** [OpenMage/magento-lts#3807] [Releases]
Hi everybody, the changelog is growing quickly so I'd like to release 20.4.0.

At the same time I think it would be great to also release 21.0 (it has to be done sometimes and some PRs are already pretty old). Maybe, since there's especially one very big change (the removal of mysql4 classes) we could have a 21.0-rc1, and have a final release in a couple of months if everything goes ok.

Is this ok?

**Comment by addison74** [comment]
I agree with the release of a new version 20.4.0.

On the subject of the main and next branches, maybe we should resume the discussions. Version 19 has been buried when we moved our work on v20, but we have to offer a stable version. It could be a serious problem, when we have to move forward with v21. Some changes will create trouble for sure. 

What is the next branch anyway? PR's were created for it, but I doubt that anyone uses it in testing, not to mention production. I remember I ran the interface once on this branch. Let's take one example, removing the Prototype library from the OpenMage code could also be done in the main branch. keeping all the JS files in place for BC. As well as for the other changes. 

Main should be for existing stores and next should be for new stores. It's just an opinion and we can have others on this topic.

**Reply by fballiano** [reply]
That discussion was months long and it doesn't seem right to rediscuss it, unless a new RFC is created by somebody.

v21 is v20 plus some PRs, it should be that unstable and it's counterproductive to never release those PRs (because the more they differ the more we'll be afraid, and that's wrong!), what did we make them for?

what's technically a BC should be in the next branch, all the rest in the main branch.

and new major should be released every once in a while (like the RFC says) like every other software, and the stores have to understand that they have to upgrade, like that they have to upgrade php and everything.

> Main should be for existing stores and next should be for new stores.

I disagree completely with this vision, every store should be on the last version of openmage. we should keeping introducing new things and detach ourselves from the past (in a healthy way obviously, not just for fun) but it makes no sense if the stores do not upgrade.

**Reply by addison74** [reply]
You have to look from the perspective of a store that will have to pay for OS update services, PHP, OpenMage, extensions, custom code. If these updates are becoming expensive, that customer will look for other platforms. There are cases where such changes are made in weeks or even months. Due to the issues in some markets, many stores have reduced their profit margins and any additional expenditure or blocking of the activity becomes a big problem.

**Reply by fballiano** [reply]
the updates we require are very well thought and shouldn't be expensive, if we don't want anybody to ever pay anything we won't progress in anything with openmage and we won't have any job in the meanwhile.

store don't have to forcefully upgrade, but we don't have to walk on crystals all the time because of that.

again, why did we discuss it for months, why did we approve those prs if we never want to release them?

**Reply by addison74** [reply]
You most likely misunderstood me. I naturally asked myself what would happen if we make Next the production version this year. There are some PRs that will create many issues. I agree that each store has a budget allocated to developers, but are these developers aware of the changes in the next branch? 

Maybe it's not my problem, but I also put myself in the place of the store owner who has no specific knowledge and is in the hands of the developer. Maybe that's why many went towards Shopify, WooCommerce, the developer decided. It would be wonderful to make as little effort as possible and earn as much money as possible. 

Because you asked me in the first post, I will give you the answer after this conversation about Next. I don't know what challenges await me, if I were to create stores from scratch with it I wouldn't worry. Let's see other opinions from those who are using the current version. Of all those who read here, how many have any idea about Next, about the changes we've made? There are quite a lot of BCs, they must update their extensions. One of the PRs is related to MySQL4 vs Resource.

**Reply by fballiano** [reply]
the changes will be in the release notes, and we can add a paragraph in the readme, same as we did with the v20 changes :-)

we already make the changes as little as possible, a sass absorbs the costs of upgrading in a different way (but the customer pays every month, and shopify plus is not cheap at all) but I know for sure that also shopify has breaking changes (I don't remember if a new version of the editor or something like that). In laravel they had to create an AI to upgrade between versions because of the many incopatibilities :-)

In Next there are ONLY breaking changes, that's the reason why that branch exists :-) the list is minimal tho and it was already put here: https://github.com/OpenMage/magento-lts/discussions/3654#discussion-5847591 but sure, it's a new major version so changes are there.

**Reply by alexh-swdev** [reply]
My main issue would be: (as a developer with limited time resources on this topic (I have so much time at hand, that our switch to Shopware 5 since the Magento 1 EOL announcement turned to a Shopware 6 migration which will turn into.. :) (Thanks for your OM-job so there is no REAL  pressure here yet ;) ) 
* Are the old and not-very-nice-too-look-at legacy themes (and rwd) still working in "Next"? Somewhere I read you want to throw them away. Over time there are many customizations to those and I am not sure how long it will take to migrate them to a new theme base.
* Are legacy plugins (not maintained by any1 other than me) still working in "Next"? (I am not necesarily talking about PHP versions... this is doable and necessary - but with the "Next" codebase)
* What das thoe MySQL4 removal mean for running shops?

These are just my concerns and I'll probably find out soon enough. But I understand that at some point, OM must move forward. The more help you can provide to devs to keep up, the better. 

And maybe this only partly on-topic but you mentioned the readme: Please (with whipped cream and cherry on to) :) Make breaking changes visible in the release anouncement. Took me a while to notice the xml-namespace change e.g. Or the other change you did in the API access with the url.

**Reply by kiatng** [reply]
Quick answers to @alexh-swdev 3 questions

- There are 2 core themes: base and rwd, the latter fallback to base. Yes, both working in `next` branch.
- Hard to say if legacy plugins would work in `next` branch. If PHP incompatibility are taken care of, I would think it's not so difficult to make it compatible with `next` branch. I have massive custom code in my projects, 99% were about PHP8.1 compatibility fixes.
- For very old code, you just need to make sure your resource model extends to `Mage_Core_Model_Resource_Db_Abstract` instead of the  mysql4 class. Once you make the changes, it should work.
